src/runtime/weak_ref_count.rs

```diff
@@ -0,0 +1,315 @@
+        let count_or_pointer = self.0.load(Ordering::Relaxed);
+
+        loop {
+            if is_weak_ref(count_or_pointer) {
+                unsafe {
+                    let tear_off = TearOff::decode(count_or_pointer);
+                    let remaining = tear_off.strong_count.release();
+
+                    // If this is the last strong reference, we can release the weak reference implied by the strong reference.
+                    // There may still be weak references, so the WeakRelease is called to handle such possibilities.
+                    if remaining == 0 {
+                        TearOff::WeakRelease(&mut tear_off.weak_vtable as *mut _ as _);
+                    }
+
+                    return remaining;
+                }
+            }
+
+            if self
+                .0
+                .compare_exchange_weak(
+                    count_or_pointer,
+                    count_or_pointer - 1,
+                    Ordering::Release,
+                    Ordering::Relaxed,
+                )
+                .is_ok()
+            {
+                return count_or_pointer as u32 - 1;
+            }
+        }
+    }
+
+    pub unsafe fn query(&self, iid: &::windows::Guid, object: RawPtr) -> RawPtr {
+        if iid != &IWeakReferenceSource::IID {
+            return std::ptr::null_mut();
+        }
+
+        let count_or_pointer = self.0.load(Ordering::Relaxed);
+
+        if is_weak_ref(count_or_pointer) {
+            return TearOff::from_encoding(count_or_pointer);
+        }
+
+        let tear_off = TearOff::new(object, count_or_pointer as _);
+        let encoding: usize =
+            ((tear_off.abi() as usize) >> 1) | (1 << (std::mem::size_of::<usize>() * 8 - 1));
+
+        loop {
+            if self
+                .0
+                .compare_exchange_weak(
+                    count_or_pointer,
+                    encoding as _,
+                    Ordering::AcqRel,
+                    Ordering::Relaxed,
+                )
+                .is_ok()
+            {
+                let result: RawPtr = std::mem::transmute(tear_off);
+                TearOff::from_strong_ptr(result).strong_count.add_ref();
+                return result;
+            }
+
+            if is_weak_ref(count_or_pointer) {
+                return TearOff::from_encoding(count_or_pointer);
+            }
+
+            TearOff::from_strong_ptr(tear_off.abi())
+                .strong_count
+                .0
+                .store(count_or_pointer as _, Ordering::SeqCst);
+        }
+    }
+}
+
+fn is_weak_ref(value: isize) -> bool {
+    value < 0
+}
+
+#[repr(C)]
+struct TearOff {
+    strong_vtable: *const IWeakReferenceSource_abi,
+    weak_vtable: *const IWeakReference_abi,
+    object: RawPtr,
+    strong_count: RefCount,
+    weak_count: RefCount,
+}
+
+impl TearOff {
+    unsafe fn new(object: RawPtr, strong_count: u32) -> IWeakReferenceSource {
+        std::mem::transmute(::std::boxed::Box::new(TearOff {
```

Comment:
Was this code generated at some point? I don't think we need the fully qualified paths here.